[settings] Add output_buildfile_suffix option

mstg opened the pull request:
Related issue: #92 

Default is BUILD
Other values can be for example: BUILD.bazel
Any value is accepted.

This is to follow the default behaviour of other bazel related tools
that generate BUILD.bazel instead of BUILD firstly to separate
what is generated and what is written and to avoid name collisions on
case insensitive file systems (Windows + WSL) since some crates can
have a directory named build which leads to cargo-raze unable to create
a file called BUILD. A project having a directory named BUILD.bazel is
much less likely or less to none chance.

How to use (example):
Add following to the `Cargo.toml` file
```
[raze]
output_buildfile_suffix = "BUILD.bazel"
```

I am still very new to Rust, and may have written something not idiomatic at all, and would appreciate any advice to make it better.

EDIT:
This PR initially added a `GenMode` called VendoredBuildBazel but upon feedback changed to another option called `output_buildfile_suffix` to ensure greater control and compatibility with all `GenMode`s

acmcarther commented:
Thanks for sending out a PR to fix this!

My general preference is to add this as a new `RazeSetting`, eg `RazeSetting.output_buildfile_suffix` rather than as a `RazeSetting.GenMode`, perhaps with a default value of "BUILD". I think this makes more sense, since the file name decision is relevant to all GenModes. Then, both branches of `CrateCatalogEntry.local_build_path` can consult the new setting when yielding a file path.

I'm definitely open to learning if there are additional benefits to having this as a full GenMode though.

mstg commented:
Hello!

First of all, thanks for your feedback! I greatly appreciate it.

I thought of that first, but since remote mode uses `package-name.BUILD` format I kind of assumed it would not be necessary for that mode specifically, but thinking back if any new `GenMode` (if needed for whatever reason) is added this option may be useful for that too. It also makes sense since this actually does not change the "storage" strategy.

I'll see if I can set something up quick.

mstg commented:
I have since added a new option called `output_buildfile_suffix` and removed the `VendorBuildBazel` GenMode.
The change was introduced in a force push.

I have also added some smoke tests to ensure the new build files work.

If anything is missing or should be done another way, I would greatly appreciate your feedback.

Thanks.

mstg commented:
There is some tests failing when `GenMode` is set to `Remote` especially with the `build_file` directive in `crates.bzl`, even though it should be working. I'll take a closer look at this tomorrow, but feel like this is the right path for now.

acmcarther commented on impl/src/planning.rs:
```diff
@@ -133,6 +135,7 @@ struct CrateSubplanner<'planner> {
 pub struct PlannedBuild {
   pub workspace_context: WorkspaceContext,
   pub crate_contexts: Vec<CrateContext>,
+  pub buildfile_suffix: String,
```
Nit: I think RenderDetails is a better place for this field. It is currently initialized [here](https://github.com/google/cargo-raze/blob/30fb8b43192858316272b3a0e79c9b2ef86ed0bb/impl/src/bin/cargo-raze.rs#L134-L136), and the suffix can be extracted from the settings at that location.

acmcarther commented on impl/src/bazel.rs:
```diff
@@ -250,39 +253,42 @@ impl BuildRenderer for BazelRenderer {
       edition: "2015".to_owned(),
       features: vec!["feature1".to_owned(), "feature2".to_owned()].to_owned(),
-      expected_build_path: "vendor/test-binary-1.1.1/BUILD".to_owned(),
+      expected_build_path: format!("vendor/test-binary-1.1.1/{}", buildfile_suffix).to_owned(),
```
You can omit this `to_owned()` now that you've switched to `format!`.

mstg commented:
Hello!

Following additional changes has been done:
* Additional smoke tests removed
* `vendored/hello_cargo_library` now generates `BUILD.bazel` files
* `buildfile_suffix` is now moved to `RenderDetails`
* Redundant `to_owned()` is now removed

I have also rebased the commits to one commit only.

If you have any other suggestions, I would greatly appreciate it.
Thanks!

acmcarther commented on smoke-test.sh:
```diff
@@ -38,6 +38,7 @@ git_repository(
     remote = "https://github.com/bazelbuild/rules_rust.git",
 )
 
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
```
Thanks for fixing this!

mstg commented:
Yes, it is the work of rustfmt :)

Awesome, thanks for bearing with me.

acmcarther commented:
Just tried to publish but I've got an ancient rustc on my workstation. Got a meeting now, but will push 0.2.3 within the hour.

mstg commented:
Great! I also forgot to add any documentation for this, so if you could do that or just ping me that would be awesome.

acmcarther commented:
Pushed 0.2.3 with this change.

I don't think there is a dedicated place for documentation of individual configuration fields beyond the link in the README to impl/settings.rs, so I guess what we have is sufficient.
